hazelcast/src/hazelcast/client/spi/impl/ClientExecutionServiceImpl.cpp

```diff
@@ -0,0 +1,201 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "hazelcast/util/IOUtil.h"
+#include <boost/foreach.hpp>
+#include "hazelcast/client/spi/impl/ClientExecutionServiceImpl.h"
+#include "hazelcast/client/ClientProperties.h"
+#include "hazelcast/util/Util.h"
+#include "hazelcast/util/Runnable.h"
+#include "hazelcast/util/Thread.h"
+#include "hazelcast/util/impl/SimpleExecutorService.h"
+
+namespace hazelcast {
+    namespace client {
+        namespace spi {
+            namespace impl {
+
+                ClientExecutionServiceImpl::ClientExecutionServiceImpl(const std::string &name,
+                                                                       const ClientProperties &clientProperties,
+                                                                       int32_t poolSize)
+                        : logger(util::ILogger::getLogger()) {
+
+                    int internalPoolSize = clientProperties.getInternalExecutorPoolSize().getInteger();
+                    if (internalPoolSize <= 0) {
+                        internalPoolSize = util::IOUtil::to_value<int>(
+                                ClientProperties::INTERNAL_EXECUTOR_POOL_SIZE_DEFAULT);
+                    }
+
+                    int executorPoolSize = poolSize;
+                    if (executorPoolSize <= 0) {
+                        executorPoolSize = util::getAvailableCoreCount();
+                    }
+
+                    internalExecutor.reset(
+                            new util::impl::SimpleExecutorService(logger, name + ".internal-", internalPoolSize,
+                                                                  INT32_MAX));
+                    userExecutor.reset(
+                            new util::impl::SimpleExecutorService(logger, name + ".user-", executorPoolSize,
+                                                                  INT32_MAX));
+                }
+
+                util::ExecutorService &ClientExecutionServiceImpl::getUserExecutor() {
+                    return *userExecutor;
+                }
+
+                void ClientExecutionServiceImpl::execute(const boost::shared_ptr<util::Runnable> &command) {
+                    internalExecutor->execute(command);
+                }
+
+                void ClientExecutionServiceImpl::shutdown() {
+                    shutdownExecutor("user", *userExecutor, logger);
+                    shutdownExecutor("internal", *internalExecutor, logger);
+
+                    boost::shared_ptr<AbstractRunner> runner;
+                    while ((runner = delayedRunners.poll()).get()) {
```

Comment:
This `delayedRunners` are only polled on shutdown. Something seems missing here ?

Reply:
delayedRunners queue is only used for keeping the reference of the runner thread so that they are not destructed.